add cpp serializer for flat tensor (ptd)

JacobSzwejbka opened the pull request:
Summary:
Leverage the flatbuffer builder apis to generate a .ptd on device. This will be used by training for checkpointing.

No other use cases really exist for generating a .ptd on device right now, so I didnt worry about making this more easily extensible by coming up with a cpp equivalent of a cord or by trying to think ahead of how this might integrate with delegates.

Later if we add support for delegates owning the weights under ET training we can revisit this.

Differential Revision: D67992901

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -0,0 +1,138 @@
+namespace flat_tensor {
+
+namespace {
+  size_t padding_required(size_t offset, size_t alignment) {
```
Code shouldn't be indented inside a namespace

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -0,0 +1,138 @@
+      out.write("\0", 1);
+    }
+  }
+}
```
```suggestion
} // namespace
```

dbort commented on extension/flat_tensor/serialize/serialize.h:
```diff
@@ -0,0 +1,67 @@
+namespace extension {
+namespace flat_tensor {
+namespace internal {
+  // cpp mirror of the python definition in serialize.py
```
Code shouldn't be indented inside a namespace

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -0,0 +1,138 @@
+
+namespace {
+  size_t padding_required(size_t offset, size_t alignment) {
+    // Returns the padding required to align `offset` to `alignment`.
```
This should be a comment above the function, not inside the function. Same for the next one

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -0,0 +1,138 @@
+  }
+}
+
+ET_EXPERIMENTAL runtime::Error save_ptd(
```
The EXPERIMENTAL annotations need to be in the .h file not the .cpp file, otherwise call sites won't emit warnings for using them

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -0,0 +1,138 @@
+  const std::map<std::string, exec_aten::Tensor>& tensor_map,
+  const size_t tensor_alignment
+) {
+  // Assert the system is little endian
```
Not really necessary because the other save_ptd will do the same when this calls it

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -0,0 +1,138 @@
+  const std::map<std::string, exec_aten::Tensor>& tensor_map,
+  const size_t tensor_alignment
+) {
+  // Assert the system is little endian
```
Please explain why you need to make this check.

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -0,0 +1,138 @@
+    tensors.push_back(tensor_metadata);      // Precalculate the size of the data blob
+    total_segment_size += aligned_size(tensor.nbytes(), tensor_alignment);
+  }
+  // Only have one segment
```
Why is there only one segment? Seems like we'd want to be able to load and free large tensors independently

JacobSzwejbka replied:
Thats the behavior of the python serializer today. I was just mimicking it. Theres not an immediate need to separate today, it will come if we start combining mutable and non mutable weights in the same .ptd for training, or with delegates presumably. Can add that functionality later though with no bc breaks.

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -0,0 +1,138 @@
+  // Only have one segment
+  buffers.push_back(::flat_tensor::CreateDataSegment(builder, 0, total_segment_size));
+  
+  auto flat_tensor = CreateFlatTensor(builder, internal::VERSION, tensor_alignment, builder.CreateVector(tensors), builder.CreateVector(buffers));
```
Longer than 80 columns, probably needs a lint reformat

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -0,0 +1,138 @@
+  auto padded_header_size = aligned_size(internal::HEADER_EXPECTED_LENGTH, tensor_alignment);
+
+  // Serialize header
+  out.write(internal::HEADER_MAGIC, sizeof(internal::HEADER_MAGIC));
```
This is arranged differently from PTE files. I hid the extended header inside the flatbuffer data so that a PTE file can still be read by standard flatbuffer tools, which has been really useful.

See https://github.com/pytorch/executorch/blob/2411c9913309914ce8dd6db7e17b96ef85907bd6/exir/_serialize/_program.py#L55 for that logic

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -0,0 +1,138 @@
+  out.write(reinterpret_cast<const char *>(&header.segment_base_offset), sizeof(header.segment_base_offset));
+  out.write(reinterpret_cast<const char *>(&header.segment_data_size), sizeof(header.segment_data_size));
+
+  // serialize header padding
```
Super nit: these steps aren't serializing, they're writing. Serializing is the act of arranging the data in a particular format, regardless of where that data lives. These steps are writing out already-serialized data.

dbort commented on extension/flat_tensor/test/test_serialize.cpp:
```diff
@@ -0,0 +1,115 @@
+  ASSERT_EQ(
+      *(uint32_t*)(byte_buffer + 4),
+      executorch::extension::FlatTensorHeader::kHeaderExpectedLength);
+  ASSERT_EQ(
+      *(uint64_t*)(byte_buffer + 8),
+      48); // Header is 40 bytes + 8 bytes of padding today, and then the
+           // flatbuffer starts.
+  ASSERT_EQ(
+      *(uint64_t*)(byte_buffer + 16),
+      224); // This is fragile, and depends on the schema, the builder, and the
+            // padding needed.
```
Please mention which fields these are. Flatbuffer offset + flatbuffer length?

dbort commented on extension/flat_tensor/test/test_serialize.cpp:
```diff
@@ -0,0 +1,115 @@
+  ASSERT_EQ(*(uint64_t*)(byte_buffer + 24), segment_offset);
+  ASSERT_EQ(
+      *(uint64_t*)(byte_buffer + 32),
+      32); // Segment total size, 8 bytes of data (2 floats), 24 bytes of
```
If there's only one segment, there shouldn't be any padding after it: there should only be padding between adjacent segments

dbort commented on extension/flat_tensor/test/test_serialize.cpp:
```diff
@@ -0,0 +1,115 @@
+           // padding.
+
+  // Check Flatbuffer
+  auto flat_tensor = ::flat_tensor::GetFlatTensor(byte_buffer + 48);
```
Once the extended header is hidden properly, you should be able to load the flatbuffer starting at offset zero of byte_buffer.

JacobSzwejbka replied:
Yeah I update the test in the following diff.

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -0,0 +1,160 @@
+        0, // segment index
+        total_segment_size);
+    tensors.push_back(
+        tensor_metadata); // Precalculate the size of the data blob
+    total_segment_size += aligned_size(tensor.nbytes(), tensor_alignment);
```
```suggestion
        tensor_metadata);
    // Precalculate the size of the data blob
    total_segment_size += aligned_size(tensor.nbytes(), tensor_alignment);
```

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -0,0 +1,160 @@
+  for (const auto& [name, tensor] : tensor_map) {
+    out.write(
+        reinterpret_cast<const char*>(tensor.data_ptr()), tensor.nbytes());
+    write_nulls(out, padding_required(tensor.nbytes(), tensor_alignment));
```
The final entry doesn't need padding after it

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -0,0 +1,160 @@
+      builder.GetSize());
+
+  // Write flatbuffer padding
+  write_nulls(out, padding_required(builder.GetSize(), tensor_alignment));
```
Really, only stuff after this point needs to be aligned to tensor_alignment; the header and flatbuffer data could just be aligned to 8. PTE aligns its flatbuffer data because the flatbuffer actually contains tensor buffers, but we don't do that in this format.

Aligning to a larger value is fine, though. And we could change the earlier alignments at any time without breaking BC

JacobSzwejbka replied:
Ah yeah Im matching the python behavior here which just padded everything to tensor alignment.

dbort reviewed (changes requested):
Thanks for the changes -- the tests look great

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -0,0 +1,173 @@
+
+    tensors.push_back(tensor_metadata);
+    // Don't pad last entry.
+    if (i != 1) {
+      // Precalculate the size of the data blob.
+      total_segment_size += aligned_size(tensor.nbytes(), tensor_alignment);
+    }
```
Missing the `--i` like in the loop below. Please add a test that would have caught these issues.

dbort commented on extension/flat_tensor/serialize/serialize.cpp:
```diff
@@ -0,0 +1,173 @@
+    // Don't pad last entry.
+    if (i != 1) {
+      // Precalculate the size of the data blob.
+      total_segment_size += aligned_size(tensor.nbytes(), tensor_alignment);
```
If there's only one tensor, total_segment_size will be zero. I think you need to increment the size by the unpadded amount for the final entry.

dbort commented on extension/flat_tensor/serialize/serialize.h:
```diff
@@ -0,0 +1,55 @@
+
+/**
+ *
+ * Schema version of the .ptd format. Should be kept in sync with Python.
+ * https://github.com/pytorch/executorch/blob/main/extension/flat_tensor/serialize/serialize.py#L269
```
The problem with pointing to a specific line is that it won't be the right line after serialize.py changes over time. I'd just point to the file in general. And since it's in the same dir you could skip the full URL

```suggestion
 * Schema version of the .ptd format. Should be kept in sync with the same value in serialize.py.
```

dbort commented on extension/flat_tensor/serialize/serialize.py:
```diff
@@ -266,7 +266,7 @@ def serialize(
         # points to all the data segments. It will be serialized to flatbuffer.
         flat_tensor = FlatTensor(
-            version=0,
+            version=0,  # Keep in sync with c++ version number in serialize.h
```
Consider adding a top-level `_SCHEMA_VERSION: int = 0` to raise visibility of this, and to make it map more clearly to the C++ equivalent.

dbort reviewed (approved):
Thanks for all of the changes -- looks great apart from the new bugs.

JacobSzwejbka committed:
add cpp serializer for flat tensor (ptd) (#7841)

Summary:

Leverage the flatbuffer builder apis to generate a .ptd on device. This will be used by training for checkpointing.

No other use cases really exist for generating a .ptd on device right now, so I didnt worry about making this more easily extensible by coming up with a cpp equivalent of a cord or by trying to think ahead of how this might integrate with delegates.

Later if we add support for delegates owning the weights under ET training we can revisit this.

Differential Revision: D67992901
